File: 54 Spiral Matrix/Spiral Matrix.js
var spiralOrder = function (matrix) {
  let left = 0,
    right = matrix[0].length;
  let top = 0,
    bottom = matrix.length;
  let array = [];

  while (left < right && top < bottom) {
    for (let i = left; i < right; i++) {
      array.push(matrix[top][i]);
    }
    top++;

    for (let i = top; i < bottom; i++) {
      array.push(matrix[i][right - 1]);
    }
    right--;

    if (!(left < right && top < bottom)) {
      break;
    }

    for (let i = right - 1; i >= left; i--) {
      array.push(matrix[bottom - 1][i]);
    }
    bottom--;

    for (let i = bottom - 1; i >= top; i--) {
      array.push(matrix[i][left]);
    }
    left++;
  }

  return array;
};

File: 54 Spiral Matrix/Spiral Matrix.py
# Python3 versiyonunu seçmeyi unutma =)
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        left, right = 0, len(matrix[0])
        top, bottom = 0, len(matrix)
        array = []
        while left < right and top < bottom:

            for i in range(left, right):
                array.append(matrix[top][i])

            top += 1

            for i in range(top, bottom):
                array.append(matrix[i][right - 1])

            right -= 1

            if not (left < right and top < bottom):
                break

            for i in range(right - 1, left - 1, -1):
                array.append(matrix[bottom - 1][i])

            bottom -= 1

            for i in range(bottom - 1, top - 1, -1):
                array.append(matrix[i][left])

            left += 1

        return array
